src/session-lock.c

```diff
@@ -0,0 +1,309 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include "common/mem.h"
+#include "labwc.h"
+
+static struct wl_listener new_lock;
+static struct wl_listener manager_destroy;
+static struct wlr_session_lock_manager_v1 *wlr_session_lock_manager;
+
+static struct server *g_server;
+
+struct session_lock_output {
+	struct wlr_scene_tree *tree;
+	struct wlr_scene_rect *background;
+	struct session_lock *lock;
+	struct output *output;
+	struct wlr_session_lock_surface_v1 *surface;
+
+	struct wl_list link; /* session_lock.outputs */
+
+	struct wl_listener destroy;
+	struct wl_listener commit;
+	struct wl_listener surface_destroy;
+	struct wl_listener surface_map;
+};
+
+static void
+focus_surface(struct session_lock *lock, struct wlr_surface *focused)
+{
+	lock->focused = focused;
+	seat_focus_surface(&g_server->seat, focused);
+}
+
+static void
+refocus_output(struct session_lock_output *output)
+{
+	/* Try to focus another session-lock surface */
+	if (output->lock->focused != output->surface->surface) {
+		return;
+	}
+
+	struct session_lock_output *iter;
+	wl_list_for_each(iter, &output->lock->session_lock_outputs, link) {
+		if (iter == output || !iter->surface) {
+			continue;
+		}
+		if (iter->surface->mapped) {
+			focus_surface(output->lock, iter->surface->surface);
+			return;
+		}
+	}
+	focus_surface(output->lock, NULL);
+}
+
+static void
+handle_surface_map(struct wl_listener *listener, void *data)
+{
+	struct session_lock_output *surf = wl_container_of(listener, surf, surface_map);
+	if (!surf->lock->focused) {
+		focus_surface(surf->lock, surf->surface->surface);
+	}
+}
+
+static void
+handle_surface_destroy(struct wl_listener *listener, void *data)
+{
+	struct session_lock_output *output =
+		wl_container_of(listener, output, surface_destroy);
+	refocus_output(output);
+
+	assert(output->surface);
+	output->surface = NULL;
+	wl_list_remove(&output->surface_destroy.link);
+	wl_list_remove(&output->surface_map.link);
+}
+
+static void
+lock_output_reconfigure(struct session_lock_output *output)
+{
+	struct wlr_box box;
+	wlr_output_layout_get_box(g_server->output_layout, output->output->wlr_output, &box);
+	wlr_scene_rect_set_size(output->background, box.width, box.height);
+	if (output->surface) {
+		wlr_session_lock_surface_v1_configure(output->surface, box.width, box.height);
+	}
+}
+
+static void
+handle_new_surface(struct wl_listener *listener, void *data)
+{
+	struct session_lock *lock = wl_container_of(listener, lock, new_surface);
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct output *output = lock_surface->output->data;
+	struct session_lock_output *lock_output;
+	wl_list_for_each(lock_output, &lock->session_lock_outputs, link) {
+		if (lock_output->output == output) {
+			goto found_lock_output;
+		}
+	}
+	wlr_log(WLR_ERROR, "fatal: new lock surface, but no output");
+	exit(EXIT_FAILURE);
```

Comment:
Just had a quick look through the diff, this is the only place I saw a potential issue.

On some systems, turning off the monitors will cause them to actually be disconnected e.g. we don't have any outputs left. Could this trigger a complete exit of labwc when being locked and then turning off the monitors?

Reply:
Good point. Have changed to just returning.

Reply:
Hm.. now I wonder if that could cause a lockscreen to not render a surface so there is no way to unlock the session once the output is back though. I guess some tests regarding that would be useful.

Reply:
It works fine.

The surface is created when the output comes back (in `session_lock_output_create()`).

I tested with the following:

```bash
#!/bin/sh

printf '%b\n' "starting..."
sleep 1

printf '%b\n' "eDP-1 off"
wlr-randr --output eDP-1 --off

sleep 2

printf '%b\n' "swaylock before"
./build/swaylock -c 222222 -d >$HOME/log.txt 2>&1 & 
printf '%b\n' "swaylock after"

sleep 2

printf '%b\n' "eDP-1 on"
wlr-randr --output eDP-1 --on

sleep 1
printf '%b\n' "complete"
```

Reply:
Thanks for confirming.